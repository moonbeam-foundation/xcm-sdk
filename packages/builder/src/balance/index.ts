export * from './BalanceBuilder';
export * from './Erc20Abi';
export * from './BalanceBuilder.interfaces';
