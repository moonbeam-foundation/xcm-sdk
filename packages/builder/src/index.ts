import '@polkadot/api-augment';

export * from './asset-min';
export * from './balance';
export * from './builder.interfaces';
export * from './contract';
export * from './extrinsic';
export * from './fee';
export * from './mrl';
export * from './types';
