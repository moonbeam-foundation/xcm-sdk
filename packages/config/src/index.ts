export * from './ConfigService';
export * from './assets';
export * from './chains';
export * from './config.utils';
export * from './mrl-configs';
export * from './types/AssetRoute';
export * from './types/MrlAssetRoute';
export * from './types/ChainRoutes';
export * from './types/MrlChainRoutes';
export * from './xcm-configs';
