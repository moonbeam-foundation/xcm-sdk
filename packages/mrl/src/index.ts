export * from './mrl';
export * from './mrl.interfaces';
