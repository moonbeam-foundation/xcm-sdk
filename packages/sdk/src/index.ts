export * from './getTransferData/getSourceData';
export * from './getTransferData/getDestinationData';
export * from './getTransferData/getTransferData.utils';
export * from './sdk';
export * from './sdk.interfaces';
export * from './services/polkadot';
export * from './services/evm';
