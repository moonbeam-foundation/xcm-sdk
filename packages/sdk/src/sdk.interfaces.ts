import type { AnyChain, AssetAmount } from '@moonbeam-network/xcm-types';
import type { Signer } from '@polkadot/api/types';
import type { IKeyringPair } from '@polkadot/types/types';
import type { WalletClient } from 'viem';

export type EvmSigner = WalletClient;

export interface Signers {
  evmSigner?: EvmSigner;
  polkadotSigner?: Signer | IKeyringPair;
}

export interface TransferData {
  destination: DestinationChainTransferData;
  getEstimate(amount: number | string): AssetAmount;
  max: AssetAmount;
  min: AssetAmount;
  source: SourceChainTransferData;
  transfer(amount: bigint | number | string, signers: Signers): Promise<string>;
}

export interface SourceChainTransferData extends ChainTransferData {
  destinationFee: AssetAmount;
  destinationFeeBalance: AssetAmount;
  feeBalance: AssetAmount;
  max: AssetAmount;
}

export interface SovereignAccountBalance {
  feeAssetBalance: bigint | undefined;
  transferAssetBalance: bigint;
}

export interface DestinationChainTransferData extends ChainTransferData {
  sovereignAccountBalances?: SovereignAccountBalance;
}

export interface ChainTransferData {
  balance: AssetAmount;
  chain: AnyChain;
  existentialDeposit?: AssetAmount;
  fee: AssetAmount;
  min: AssetAmount;
}
