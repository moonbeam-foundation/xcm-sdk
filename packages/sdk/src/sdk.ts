import { ConfigService, xcmRoutesMap } from '@moonbeam-network/xcm-config';
import {
  type AnyAsset,
  type AnyChain,
  type AnyParachain,
  type AssetAmount,
  type Ecosystem,
  EvmParachain,
} from '@moonbeam-network/xcm-types';
import { getAssetsBalances } from './getTransferData/getSourceData';
import { getTransferData } from './getTransferData/getTransferData';
import type { TransferData } from './sdk.interfaces';

const DEFAULT_SERVICE = new ConfigService({ routes: xcmRoutesMap });

export interface SdkOptions {
  configService?: ConfigService;
  ecosystem?: Ecosystem;
}

export function Sdk({ configService, ecosystem }: SdkOptions = {}) {
  const service = configService ?? DEFAULT_SERVICE;
  const assets = service.getEcosystemAssets(ecosystem);

  return {
    assets,
    setAsset(asset: string | AnyAsset) {
      const sources = service.getSourceChains({ asset, ecosystem });

      return {
        sources,
        setSource(source: string | AnyChain) {
          const destinations = service.getDestinationChains({
            asset,
            source,
          });

          return {
            destinations,
            setDestination(destination: string | AnyChain) {
              const route = service.getAssetRoute({
                asset,
                source,
                destination,
              });

              return {
                setAddresses({
                  sourceAddress,
                  destinationAddress,
                }: {
                  sourceAddress: string;
                  destinationAddress: string;
                }): Promise<TransferData> {
                  const sourceChain = service.getChain(source);

                  if (!EvmParachain.isAnyParachain(sourceChain)) {
                    throw new Error(
                      'Source chain should be a Parachain or EvmParachain',
                    );
                  }

                  if (!EvmParachain.isAnyParachain(route.destination.chain)) {
                    throw new Error(
                      'Destination chain should be a Parachain or EvmParachain',
                    );
                  }

                  return getTransferData({
                    route,
                    sourceAddress,
                    destinationAddress,
                  });
                },
              };
            },
          };
        },
      };
    },
  };
}

export async function getParachainBalances(
  chain: AnyParachain,
  address: string,
  service: ConfigService = DEFAULT_SERVICE,
): Promise<AssetAmount[]> {
  const routes = service.getChainRoutes(chain).getRoutes();
  const balances = await getAssetsBalances({
    chain,
    routes,
    address,
  });

  return balances;
}
