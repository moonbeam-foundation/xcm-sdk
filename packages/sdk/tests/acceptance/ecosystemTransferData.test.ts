import {
  ConfigService,
  crossEcosystemsRoutesMap,
  lama,
  moonlama,
  moonsama,
} from '@moonbeam-network/xcm-config';
import type { AnyParachain, Asset } from '@moonbeam-network/xcm-types';
import { describe, expect, it } from 'vitest';
import { Sdk, type TransferData } from '../../src';
import { moonEvmAddress } from './constants';

const transferDateTestConfig: {
  asset: Asset;
  source: AnyParachain;
  destination: AnyParachain;
  sourceAddress: string;
  destinationAddress: string;
}[] = [
  {
    asset: lama,
    source: moonlama,
    sourceAddress: moonEvmAddress,
    destination: moonsama,
    destinationAddress: moonEvmAddress,
  },
];

describe('sdk/transferData ecosystem', () => {
  const configService = new ConfigService({
    routes: crossEcosystemsRoutesMap,
  });

  describe.each(transferDateTestConfig)(
    'asset $asset.key from source $source.name to destination $destination.name',
    ({ asset, source, sourceAddress, destination, destinationAddress }) => {
      it('should get expected balances', async () => {
        const transferData: TransferData = await Sdk({
          configService,
        })
          .setAsset(asset)
          .setSource(source)
          .setDestination(destination)
          .setAddresses({
            sourceAddress: sourceAddress,
            destinationAddress: destinationAddress,
          });

        expect(transferData).toMatchSnapshot();
      });
    },
  );
});
