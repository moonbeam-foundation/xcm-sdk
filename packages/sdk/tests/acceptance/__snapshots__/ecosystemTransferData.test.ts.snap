// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sdk/transferData ecosystem > asset 'lama-glmr' from source 'Moonlama' to destination 'Moonsama' > should get expected balances 1`] = `
{
  "destination": {
    "balance": _AssetAmount {
      "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "movrsama" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "movrsama",
          "min": undefined,
          "originSymbol": "MOVR-SAMA",
          "symbol": undefined,
        },
        "lama-glmr" => _ChainAsset {
          "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
          "decimals": 6,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonsama-relay",
      "explorer": undefined,
      "genesisHash": "0x13df313fb1b9b0c99dd215ff130c6437464706eabf0f63123d4da26019e1a0cc",
      "id": 1285,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonsama",
      "name": "Moonsama",
      "parachainId": 2023,
      "relayGenesisHash": "0xe553c7f33ab7dbbd8770a82169ac9ad7f49f53204e6d72d3a75f9df5d22e1183",
      "rpc": "https://viro-moonsama-rpc-1.rv.moondev.network",
      "ss58Format": 1285,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://viro-moonsama-rpc-1.rv.moondev.network/",
      ],
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "movrsama",
      "min": undefined,
      "originSymbol": "MOVR-SAMA",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
      "amount": 100000000000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "sovereignAccountBalances": undefined,
  },
  "getEstimate": [Function],
  "max": _AssetAmount {
    "address": "0x0000000000000000000000000000000000000802",
    "amount": 0n,
    "decimals": 18,
    "ids": {
      "palletInstance": 3,
    },
    "key": "lama-glmr",
    "min": undefined,
    "originSymbol": "lamaGLMR",
    "symbol": undefined,
  },
  "min": _AssetAmount {
    "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
    "amount": 100000000000000n,
    "decimals": 18,
    "ids": {
      "palletInstance": 3,
    },
    "key": "lama-glmr",
    "min": undefined,
    "originSymbol": "lamaGLMR",
    "symbol": undefined,
  },
  "source": {
    "balance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "lama-glmr" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
          "decimals": 18,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
          "decimals": 6,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonlama-relay",
      "explorer": undefined,
      "genesisHash": "0xb7e61bf426fa70533e2cc8d17bc00f1d373f3f45a01eac924da95214930a0d0e",
      "id": 1283,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonlama",
      "name": "Moonlama",
      "parachainId": 2004,
      "relayGenesisHash": "0xd0a2c75ad080394edaaa9cd1c766b543478efb415db1b6cd7772536c9a413167",
      "rpc": "https://moonlama.api.moondev.network",
      "ss58Format": 1284,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://moonlama.api.moondev.network",
      ],
    },
    "destinationFee": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 100000000000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "destinationFeeBalance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "feeBalance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "max": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
  },
  "transfer": [Function],
}
`;

exports[`sdk/transferData ecosystem > asset 'pizza' from source 'Moonlama' to destination 'Moonsama' > should get expected balances 1`] = `
{
  "destination": {
    "balance": _AssetAmount {
      "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "movrsama" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "movrsama",
          "min": undefined,
          "originSymbol": "MOVR-SAMA",
          "symbol": undefined,
        },
        "lama-glmr" => _ChainAsset {
          "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
          "decimals": 6,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonsama-relay",
      "explorer": undefined,
      "genesisHash": "0x13df313fb1b9b0c99dd215ff130c6437464706eabf0f63123d4da26019e1a0cc",
      "id": 1285,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonsama",
      "name": "Moonsama",
      "parachainId": 2023,
      "relayGenesisHash": "0xe553c7f33ab7dbbd8770a82169ac9ad7f49f53204e6d72d3a75f9df5d22e1183",
      "rpc": "https://viro-moonsama-rpc-1.rv.moondev.network",
      "ss58Format": 1285,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://viro-moonsama-rpc-1.rv.moondev.network/",
      ],
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "movrsama",
      "min": undefined,
      "originSymbol": "MOVR-SAMA",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
      "amount": 100000000000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "sovereignAccountBalances": undefined,
  },
  "getEstimate": [Function],
  "max": _AssetAmount {
    "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
    "amount": 0n,
    "decimals": 18,
    "ids": {
      "palletInstance": 110,
    },
    "key": "pizza",
    "min": undefined,
    "originSymbol": "PIZZA",
    "symbol": undefined,
  },
  "min": _AssetAmount {
    "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
    "amount": 100000000000000n,
    "decimals": 18,
    "ids": {
      "palletInstance": 3,
    },
    "key": "pizza",
    "min": undefined,
    "originSymbol": "PIZZA",
    "symbol": undefined,
  },
  "source": {
    "balance": _AssetAmount {
      "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "lama-glmr" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
          "decimals": 18,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
          "decimals": 6,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonlama-relay",
      "explorer": undefined,
      "genesisHash": "0xb7e61bf426fa70533e2cc8d17bc00f1d373f3f45a01eac924da95214930a0d0e",
      "id": 1283,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonlama",
      "name": "Moonlama",
      "parachainId": 2004,
      "relayGenesisHash": "0xd0a2c75ad080394edaaa9cd1c766b543478efb415db1b6cd7772536c9a413167",
      "rpc": "https://moonlama.api.moondev.network",
      "ss58Format": 1284,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://moonlama.api.moondev.network",
      ],
    },
    "destinationFee": _AssetAmount {
      "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
      "amount": 100000000000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "destinationFeeBalance": _AssetAmount {
      "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "feeBalance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "max": _AssetAmount {
      "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza",
      "min": undefined,
      "originSymbol": "PIZZA",
      "symbol": undefined,
    },
  },
  "transfer": [Function],
}
`;

exports[`sdk/transferData ecosystem > asset 'pizza-usdc' from source 'Moonlama' to destination 'Moonsama' > should get expected balances 1`] = `
{
  "destination": {
    "balance": _AssetAmount {
      "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "movrsama" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "movrsama",
          "min": undefined,
          "originSymbol": "MOVR-SAMA",
          "symbol": undefined,
        },
        "lama-glmr" => _ChainAsset {
          "address": "0xffffffff1a49463978f19dfd6983f2fa1885c254",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xffffffff57029d0ec61f38481147f4e1a0d5e3ed",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
          "decimals": 6,
          "ids": {
            "palletInstance": 3,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonsama-relay",
      "explorer": undefined,
      "genesisHash": "0x13df313fb1b9b0c99dd215ff130c6437464706eabf0f63123d4da26019e1a0cc",
      "id": 1285,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonsama",
      "name": "Moonsama",
      "parachainId": 2023,
      "relayGenesisHash": "0xe553c7f33ab7dbbd8770a82169ac9ad7f49f53204e6d72d3a75f9df5d22e1183",
      "rpc": "https://viro-moonsama-rpc-1.rv.moondev.network",
      "ss58Format": 1285,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://viro-moonsama-rpc-1.rv.moondev.network/",
      ],
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "movrsama",
      "min": undefined,
      "originSymbol": "MOVR-SAMA",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
      "amount": 100n,
      "decimals": 6,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 3,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "sovereignAccountBalances": undefined,
  },
  "getEstimate": [Function],
  "max": _AssetAmount {
    "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
    "amount": 0n,
    "decimals": 6,
    "ids": {
      "palletInstance": 110,
    },
    "key": "pizza-usdc",
    "min": undefined,
    "originSymbol": "pizzaUSDC",
    "symbol": undefined,
  },
  "min": _AssetAmount {
    "address": "0xffffffffa993c0ca4b46667277c3caaa5a4da783",
    "amount": 100n,
    "decimals": 6,
    "ids": {
      "palletInstance": 3,
    },
    "key": "pizza-usdc",
    "min": undefined,
    "originSymbol": "pizzaUSDC",
    "symbol": undefined,
  },
  "source": {
    "balance": _AssetAmount {
      "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "lama-glmr" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "lama-glmr",
          "min": undefined,
          "originSymbol": "lamaGLMR",
          "symbol": undefined,
        },
        "pizza" => _ChainAsset {
          "address": "0xbdd558f3be2807fce11b5fb2b2fa86f310f67855",
          "decimals": 18,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza",
          "min": undefined,
          "originSymbol": "PIZZA",
          "symbol": undefined,
        },
        "pizza-usdc" => _ChainAsset {
          "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
          "decimals": 6,
          "ids": {
            "palletInstance": 110,
          },
          "key": "pizza-usdc",
          "min": undefined,
          "originSymbol": "pizzaUSDC",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "moonlama-relay",
      "explorer": undefined,
      "genesisHash": "0xb7e61bf426fa70533e2cc8d17bc00f1d373f3f45a01eac924da95214930a0d0e",
      "id": 1283,
      "isEvmSigner": false,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonlama",
      "name": "Moonlama",
      "parachainId": 2004,
      "relayGenesisHash": "0xd0a2c75ad080394edaaa9cd1c766b543478efb415db1b6cd7772536c9a413167",
      "rpc": "https://moonlama.api.moondev.network",
      "ss58Format": 1284,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://moonlama.api.moondev.network",
      ],
    },
    "destinationFee": _AssetAmount {
      "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
      "amount": 100n,
      "decimals": 6,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "destinationFeeBalance": _AssetAmount {
      "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "feeBalance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "lama-glmr",
      "min": undefined,
      "originSymbol": "lamaGLMR",
      "symbol": undefined,
    },
    "max": _AssetAmount {
      "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0x385c00353cff4c3129306e0a6dcd1947e8354133",
      "amount": 0n,
      "decimals": 6,
      "ids": {
        "palletInstance": 110,
      },
      "key": "pizza-usdc",
      "min": undefined,
      "originSymbol": "pizzaUSDC",
      "symbol": undefined,
    },
  },
  "transfer": [Function],
}
`;
